src/base/utils/tar.h

```diff
@@ -0,0 +1,113 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2020  sledgehammer999 <hammered999@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <QByteArray>
+#include <QString>
+
+namespace Utils
+{
+    namespace Tar
+    {
+        QHash<QString, QByteArray> untar(const QByteArray &data, QString &errorString);
+
+        class StreamReader
+        {
+        public:
+            enum class Type : unsigned char
+            {
+                RegFileNull = '\0',
+                RegFileAscii = '0',
+                // We don't care about the other file types
+
+                Unknown = std::numeric_limits<unsigned char>::max()
+            };
+
+            explicit StreamReader(const QByteArray &data);
+            explicit StreamReader(const QByteArray &&data) = delete;
+
+            bool readNext();
+            QString fileName() const;
+            QByteArray fileData() const;
+            Type fileType() const;
+
+            bool atEnd() const { return m_end; }
```

Comment:
And IIRC you were the one that advocated that such simple instructions can be implemented in headers. And I think there is actually used in some headers now. Am I wrong?

Reply:
>you were the one that advocated that such simple instructions can be implemented in headers.

Yes. And I still didn't change my mind. But I don't use it since it's forbidden by qBittorrent Coding Guidelines. Maybe it's time to reconsider it?

Reply:
Sorry. I had the wrong idea in my head. Apparently we only do member variable initializations in the header. I don't know if this is constrained to only structs and very simple classes.

>Maybe it's time to reconsider it?

I don't know. I don't even remember my exact reasoning for being against this. The only thought I have for being against it now: Consistency in a class. When you're looking at a .cpp you probably want to see all member function implementations there and not go hunting them down in the header. I admit this isn't a strong case, and personally I don't have a strong position on this.

Reply:
>Consistency in a class. When you're looking at a .cpp you probably want to see all member function implementations there and not go hunting them down in the header. 

Also it is easier to read the function signatures (in header) without the function bodies.

Reply:
Okay, whatever you want... I'm used to it.